suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java

```diff
@@ -105,34 +119,155 @@ private void setServoReporter() {
         props.put(ServoReporter.class.getName(), ServoReporter.class);
     }
 
+    private AtomicLong queuedRecords = new AtomicLong(0);
+    private AtomicLong sentRecords = new AtomicLong(0);
+    private AtomicLong droppedRecords = new AtomicLong(0);
+    private volatile Action3 recordCounterListener;
+    public void setRecordCounterListener(Action3 action) {
+        this.recordCounterListener = action;
+    }
+
+    private ConcurrentSkipListSet<String> metadataFetchedTopicSet = new ConcurrentSkipListSet<>();
+    private PublishSubject<MessageContainer> stream = PublishSubject.create();
+    private Subscription subscription;
+    private ExecutorService executor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setDaemon(false).setNameFormat("kafka-metadata-fetcher-%d").build());
+
     @Override
-    public void writeTo(MessageContainer message) {
+    public void writeTo(final MessageContainer message) {
+        queuedRecords.incrementAndGet();
+        runRecordCounterListener();
+
+        if (metadataFetchedTopicSet.contains(message.getRoutingKey())) {
```

Comment:
I tested auto-topic-creation behavior by commenting the following lines in unit testing code and it worked well. So, we don't need to worry about auto-topic-creation.

```
//        TopicCommand.createTopic(zk.getZkClient(),
//            new TopicCommand.TopicCommandOptions(new String[]{
//                "--zookeeper", "dummy", "--create", "--topic", topicName1,
//                "--replication-factor", "2", "--partitions", "1"}));
//        TopicCommand.createTopic(zk.getZkClient(),
//            new TopicCommand.TopicCommandOptions(new String[]{
//                "--zookeeper", "dummy", "--create", "--topic", topicName2,
//                "--replication-factor", "2", "--partitions", "1"}));
```